File: 0239_sliding_window_maximum/max_heap/clojure/src/solution.clj
(ns solution)

(defn max-heap []
  (java.util.PriorityQueue.
   11
   (reify java.util.Comparator
     (compare [_ a b]
       (let [[va ia] a
             [vb ib] b]
         (if (= va vb)
           (compare ib ia)
           (compare vb va)))))))

(defn discard-stale! [pq left-bound]
  (loop []
    (when-let [[_ idx] (.peek pq)]
      (when (<= idx left-bound)
        (.poll pq)
        (recur)))))

(defn max-sliding-window [nums k]
  (let [pq (max-heap)]
    (loop [i 0
           result []]
      (if (= i (count nums))
        result
        (do
          (.offer pq [(nth nums i) i])
          (discard-stale! pq (- i k))
          (recur (inc i)
                 (if (>= i (dec k))
                   (conj result (first (.peek pq)))
                   result)))))))

File: 0239_sliding_window_maximum/max_heap/kotlin/solution.kt
import java.util.PriorityQueue

fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
    val heap = PriorityQueue<IntArray>(
        compareByDescending<IntArray> { it[0] }.thenBy { it[1] },
    )
    val result = mutableListOf<Int>()

    for ((i, value) in nums.withIndex()) {
        heap.add(intArrayOf(value, i))

        while (heap.peek()[1] <= i - k) {
            heap.poll()
        }

        if (i >= k - 1) {
            result.add(heap.peek()[0])
        }
    }

    return result.toIntArray()
}
